### PR TITLE
[new release] ppxlib (0.26.0)

### PR DESCRIPTION
**Pull request by pitag-ha**
Standard library for ppx rewriters

- Project page: <a href="https://github.com/ocaml-ppx/ppxlib">https://github.com/ocaml-ppx/ppxlib</a>
- Documentation: <a href="https://ocaml-ppx.github.io/ppxlib/">https://ocaml-ppx.github.io/ppxlib/</a>

##### CHANGES:

- Bump ppxlib's AST to 4.14/5.00 (ocaml-ppx/ppxlib#320, @pitag-ha)

### CODE REVIEW

**Comment by pitag-ha** [comment]
Hello, this AST bump breaks more packages than usual, so we should have a careful look at the CI. I've added a ppxlib constraint to the 9 affected packages in the ppx-universe I've created. I also have patch PRs ready for them. Like always, there are some packages that can't go into the ppx-universe though: for example, [these ones](https://github.com/ocaml-ppx/ppx_universe/blob/4.14-bump/dev/every_step.sh#L59-L77) and all packages incompatible with dune.

File: packages/traverse/traverse.0.3.0/opam [hunk 1]
```diff
@@ -16,6 +16,7 @@ depends: [
   "metapp" {>= "0.3.0"}
   "metaquot" {>= "0.3.0"}
   "odoc" {with-doc & >= "1.5.1"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/traverse/traverse.0.3.0/opam:19]: It is not traverse that fails, but metapp:
```
#=== ERROR while compiling metapp.0.4.3 =======================================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.13.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.13/.opam-switch/build/metapp.0.4.3
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p metapp -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/metapp-19-07576f.env
# output-file          ~/.opam/log/metapp-19-07576f.out
### output ###
# (cd _build/default && .ppx/e7c203cacf1cd3bb02e6242329effa31/ppx.exe --cookie 'library-name="metapp"' -o metapp/metapp.pp.mli --intf metapp/metapp.mli -corrected-suffix .ppx-corrected -diff-cmd - -dump-ast)
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/opam/.opam/4.13/lib/ocaml, /home/opam/.opam/4.13/lib/ocaml/compiler-libs
# File "_none_", line 1:
# Warning 58 [no-cmx-file]: no cmx file was found in path for module Metapp_api, and its interface was not compiled with -opaque
# (cd _build/default && .ppx/e7c203cacf1cd3bb02e6242329effa31/ppx.exe --cookie 'library-name="metapp"' -o metapp/metapp.pp.ml --impl metapp/metapp.ml -corrected-suffix .ppx-corrected -diff-cmd - -dump-ast)
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/opam/.opam/4.13/lib/ocaml, /home/opam/.opam/4.13/lib/ocaml/compiler-libs
# File "_none_", line 1:
# Warning 58 [no-cmx-file]: no cmx file was found in path for module Metapp_api, and its interface was not compiled with -opaque
# File "_none_", line 1:
# Warning 58 [no-cmx-file]: no cmx file was found in path for module Metapp_preutils, and its interface was not compiled with -opaque
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -open Stdcompat -w +32+34-40 -g -I metapp/.metapp.objs/byte -I metapp/.metapp.objs/native -I /home/opam/.opam/4.13/lib/ocaml-compiler-libs/common -I /home/opam/.opam/4.13/lib/ocaml-compiler-libs/shadow -I /home/opam/.opam/4.13/lib/ocaml/compiler-libs -I /home/opam/.opam/4.13/lib/ppx_derivers -I /home/opam/.opam/4.13/lib/ppxlib -I /home/opam/.opam/4.13/lib/ppxlib/ast -I /home/opam/.opam/4.13/lib/ppxlib/astlib -I /home/opam/.opam/4.13/lib/ppxlib/print_diff -I /home/opam/.opam/4.13/lib/ppxlib/stdppx -I /home/opam/.opam/4.13/lib/ppxlib/traverse_builtins -I /home/opam/.opam/4.13/lib/sexplib0 -I /home/opam/.opam/4.13/lib/stdcompat -I /home/opam/.opam/4.13/lib/stdlib-shims -I preutils/.metapp_preutils.objs/byte -I preutils/.metapp_preutils.objs/native -intf-suffix .ml -no-alias-deps -o metapp/.metapp.objs/native/metapp.cmx -c -impl metapp/metapp.pp.ml)
# File "metapp/metapp.ml", line 323, characters 23-26:
# 323 |           if check_pat arg then
#                              ^^^
# Error: This expression has type
#          string Ppxlib__.Import.Ast.loc list * Ppxlib__.Import.Ast.pattern
#        but an expression was expected of type
#          Ppxlib.pattern = Astlib.Ast_414.Parsetree.pattern
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlc.opt -open Stdcompat -w +32+34-40 -g -bin-annot -I metapp/.metapp.objs/byte -I /home/opam/.opam/4.13/lib/ocaml-compiler-libs/common -I /home/opam/.opam/4.13/lib/ocaml-compiler-libs/shadow -I /home/opam/.opam/4.13/lib/ocaml/compiler-libs -I /home/opam/.opam/4.13/lib/ppx_derivers -I /home/opam/.opam/4.13/lib/ppxlib -I /home/opam/.opam/4.13/lib/ppxlib/ast -I /home/opam/.opam/4.13/lib/ppxlib/astlib -I /home/opam/.opam/4.13/lib/ppxlib/print_diff -I /home/opam/.opam/4.13/lib/ppxlib/stdppx -I /home/opam/.opam/4.13/lib/ppxlib/traverse_builtins -I /home/opam/.opam/4.13/lib/sexplib0 -I /home/opam/.opam/4.13/lib/stdcompat -I /home/opam/.opam/4.13/lib/stdlib-shims -I preutils/.metapp_preutils.objs/byte -intf-suffix .ml -no-alias-deps -o metapp/.metapp.objs/byte/metapp.cmo -c -impl metapp/metapp.pp.ml)
# File "metapp/metapp.ml", line 323, characters 23-26:
# 323 |           if check_pat arg then
#                              ^^^
# Error: This expression has type
#          string Ppxlib__.Import.Ast.loc list * Ppxlib__.Import.Ast.pattern
#        but an expression was expected of type
#          Ppxlib.pattern = Astlib.Ast_414.Parsetree.pattern
```
This should not be needed

File: packages/clangml-transforms/clangml-transforms.0.25/opam [hunk 2]
```diff
@@ -13,6 +13,7 @@ depends: [
   "dolog" {>= "4.0.0" & < "5.0.0"}
   "traverse" {>= "0.2.0"}
   "refl" {>= "0.1.0"}
+  "ppxlib" {< "0.26.0"}
 ]
```
Comment (mseri) [inline packages/clangml-transforms/clangml-transforms.0.25/opam:17]: This fails due to metapp failing (a dependency of traverse). There should be no dependency on ppxlib here.
See https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/a8aa66bb463131a112f5b73f2430e04ec04d82f2/variant/opam-2.1,compilers,4.13,ppxlib.0.26.0,revdeps,clangml-transforms.0.26

File: packages/traverse/traverse.0.3.0/opam (hunk 1 above)
Comment (pitag-ha) [reply packages/traverse/traverse.0.3.0/opam:19]: Ooh, cool, that's probably the same for a couple of other packages as well. Thanks!

File: packages/memcad/memcad.1.1.0/opam [hunk 3]
```diff
@@ -33,6 +33,7 @@ depends: [
   "qtest"
   "mlbdd" {>= "0.5"}
   "setr" {>= "0.1.1"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/memcad/memcad.1.1.0/opam:36]: This also should not be here. The CI failure is due to metapp 0.4.3

File: packages/traverse/traverse.0.3.0/opam (hunk 1 above)
Comment (mseri) [reply packages/traverse/traverse.0.3.0/opam:19]: Lookign at the logs, I think all of

- [metapp.0.4.0 (failed: This expression has type)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/a8aa66bb463131a112f5b73f2430e04ec04d82f2/variant/opam-2.1,compilers,4.13,ppxlib.0.26.0,revdeps,metapp.0.4.0)
- [metapp.0.4.1 (failed: This expression has type)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/a8aa66bb463131a112f5b73f2430e04ec04d82f2/variant/opam-2.1,compilers,4.13,ppxlib.0.26.0,revdeps,metapp.0.4.1)
- [metapp.0.4.2 (failed: This expression has type)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/a8aa66bb463131a112f5b73f2430e04ec04d82f2/variant/opam-2.1,compilers,4.13,ppxlib.0.26.0,revdeps,metapp.0.4.2)
- [metapp.0.4.3 (failed: This expression has type)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/a8aa66bb463131a112f5b73f2430e04ec04d82f2/variant/opam-2.1,compilers,4.13,ppxlib.0.26.0,revdeps,metapp.0.4.3)

need the upper bound and will avoid a number of failures

File: packages/metaquot/metaquot.0.5.0/opam [hunk 4]
```diff
@@ -12,7 +12,7 @@ bug-reports: "https://github.com/thierry-martinez/metaquot"
 depends: [
   "ocaml" {>= "4.08.0" & < "4.14"}
   "stdcompat" {>= "12"}
-  "ppxlib" {>= "0.22.0"}
+  "ppxlib" {>= "0.22.0" & < "0.26.0"}
```
Comment (mseri) [inline packages/metaquot/metaquot.0.5.0/opam:15]: Currently fails due to metapp. I would omit this one for the moment

File: packages/clangml/clangml.4.5.0/opam [hunk 5]
```diff
@@ -23,7 +23,7 @@ depends: [
   "metaquot" {>= "0.4.0"}
   "refl" {>= "0.4.0"}
   "odoc" {with-doc & >= "1.5.1"}
-  "ppxlib" {>= "0.23"}
+  "ppxlib" {>= "0.23" & < "0.26.0"}
```
Comment (mseri) [inline packages/clangml/clangml.4.5.0/opam:26]: This is actually a metapp failure for now. I think we should remove this upper bound for now

File: packages/metaquot/metaquot.0.5.1/opam [hunk 6]
```diff
@@ -12,7 +12,7 @@ bug-reports: "https://github.com/thierry-martinez/metaquot"
 depends: [
   "ocaml" {>= "4.08.0" & < "4.14"}
   "stdcompat" {>= "12"}
-  "ppxlib" {>= "0.22.0"}
+  "ppxlib" {>= "0.22.0" & < "0.26.0"}
   "ocamlfind" {>= "1.8.1"}
```
Comment (mseri) [inline packages/metaquot/metaquot.0.5.1/opam:16]: Currently fails due to metapp. I would omit this one for the moment

File: packages/override/override.0.4.0/opam [hunk 7]
```diff
@@ -13,7 +13,7 @@ bug-reports: "https://github.com/thierry-martinez/override"
 depends: [
   "dune" {>= "2.9"}
   "ocaml" {>= "4.08.0"}
-  "ppxlib" {>= "0.23.0"}
+  "ppxlib" {>= "0.23.0" & < "0.26.0"}
```
Comment (mseri) [inline packages/override/override.0.4.0/opam:16]: Currently fails due to metapp. I would omit this one for the moment

File: packages/pattern/pattern.0.3.0/opam [hunk 8]
```diff
@@ -26,6 +26,7 @@ depends: [
   "refl" {>= "0.3.0"}
   "ocaml-migrate-parsetree" {>= "1.5.0"}
   "stdcompat" {>= "10"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/pattern/pattern.0.3.0/opam:29]: Currently fails due to metapp. I would omit this one for the moment

File: packages/pattern/pattern.0.3.1/opam [hunk 9]
```diff
@@ -26,6 +26,7 @@ depends: [
   "refl" {>= "0.3.0"}
   "ocaml-migrate-parsetree" {>= "1.5.0"}
   "stdcompat" {>= "10"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/pattern/pattern.0.3.1/opam:29]: Currently fails due to metapp. I would omit this one for the moment

File: packages/ppx_pbt/ppx_pbt.0.1.0/opam [hunk 10]
```diff
@@ -19,6 +19,7 @@ depends: [
   "menhir" {>= "20180523"}
   "qcheck" {>= "0.17"}
   "metaquot" {>= "0.4"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/ppx_pbt/ppx_pbt.0.1.0/opam:22]: Currently fails due to metapp. I would omit this one for the moment

File: packages/pyast/pyast.0.1.0/opam [hunk 11]
```diff
@@ -23,6 +23,7 @@ depends: [
   "alcotest" {with-test & >= "1.4.0"}
   "conf-python-3-dev" {with-test}
   "odoc" {with-doc & >= "1.5.1"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/pyast/pyast.0.1.0/opam:26]: Currently fails due to metapp. I would omit this one for the moment

File: packages/pyast/pyast.0.1.1/opam [hunk 12]
```diff
@@ -23,6 +23,7 @@ depends: [
   "alcotest" {with-test & >= "1.4.0"}
   "conf-python-3-dev" {with-test}
   "odoc" {with-doc & >= "1.5.1"}
+  "ppxlib" {< "0.26.0"}
 ]
 build: [
```
Comment (mseri) [inline packages/pyast/pyast.0.1.1/opam:28]: Currently fails due to metapp. I would omit this one for the moment

File: packages/refl/refl.0.1.0/opam [hunk 13]
```diff
@@ -17,6 +17,7 @@ depends: [
   "metapp" {>= "0.2.0" & < "0.3.0"}
   "metaquot" {>= "0.2.0" & < "0.3.0"}
   "traverse" {>= "0.2.0" & < "0.3.0"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/refl/refl.0.1.0/opam:20]: Currently fails due to metapp. I would omit this one for the moment

File: packages/refl/refl.0.3.0/opam [hunk 14]
```diff
@@ -18,6 +18,7 @@ depends: [
   "metaquot" {>= "0.3.0" & < "0.4.0"}
   "traverse" {>= "0.3.0"}
   "odoc" {with-doc & >= "1.5.1"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/refl/refl.0.3.0/opam:21]: Currently fails due to metapp. I would omit this one for the moment

File: packages/refl/refl.0.4.0/opam [hunk 15]
```diff
@@ -18,6 +18,7 @@ depends: [
   "metaquot" {>= "0.4.0"}
   "traverse" {>= "0.3.0"}
   "odoc" {with-doc & >= "1.5.1"}
+  "ppxlib" {< "0.26.0"}
```
Comment (mseri) [inline packages/refl/refl.0.4.0/opam:21]: Currently fails due to metapp. I would omit this one for the moment

**Comment by mseri** [comment]
Looking at [cc1a40b](https://github.com/ocaml/opam-repository/pull/20977/commits/cc1a40b931d77d251efa7883ad4ddd1543d01224), you should only have the bounds on metapp and the gen_js_api tests (for now). All the other packages are failing due to metapp failing

**Comment by pitag-ha** [comment]
Thanks for noticing @mseri! Cool that in the end, the impact of this release is less than a minute ago I thought it was.

There's still at least one package that I'll probably need to look into: `ppx_yojson`. It seems like the tests in there might be failing now due to a change in `Ppxlib.Pprintast`. I'll look into that tomorrow (unless you think the failure is unrelated to this release).

**Comment by mseri** [comment]
Thanks for taking care of the revdeps!

**Comment by mseri** [comment]
`ppx_yojson` failures are just expect-tests. We could leave it as is, or disable the tests with recent ppxlib `{with-test &  ppxlib:version < "0.26.0"}`

**Comment by mseri** [comment]
I'll fix those remaining failures separately

**Pull request merged**